File: src/tests/Behavioral/GenericStructEquality/GenericStructEquality.go
package main

import "fmt"

// The unique.Handle shape: the only field is *T, whose pointer-identity == is
// valid for every T — equality must not depend on T's own constraints.
type handle[T comparable] struct {
	p *T
}

// A plain type-parameter field: no == is available for T itself, so equality
// falls back to the field type's own equality.
type box[T any] struct {
	v T
}

// No field mentions T at all (the internal/weak.Pointer / Cache[T] shape).
type meta[T any] struct {
	hits int
}

// Mixed: a type-parameter field beside a concrete field (the database/sql.Null shape).
type optval[T any] struct {
	v     T
	valid bool
}

// A generic struct field inside another generic struct.
type outer[T comparable] struct {
	inner box[T]
	n     int
}

func main() {
	x, y := 5, 5
	h1, h2, h3 := handle[int]{&x}, handle[int]{&x}, handle[int]{&y}
	fmt.Println(h1 == h2, h1 == h3, h1 == handle[int]{})

	b1 := box[string]{"go"}
	fmt.Println(b1 == box[string]{"go"}, b1 == box[string]{"cs"})
	fmt.Println(box[int]{7} == box[int]{7}, box[int]{7} == box[int]{8})

	m1, m2 := meta[string]{3}, meta[float64]{3}
	fmt.Println(m1 == meta[string]{3}, m2 == meta[float64]{4})

	n1, n2 := optval[string]{"a", true}, optval[string]{"a", false}
	fmt.Println(n1 == optval[string]{"a", true}, n1 == n2)

	o1, o2 := outer[string]{box[string]{"k"}, 1}, outer[string]{box[string]{"k"}, 1}
	fmt.Println(o1 == o2, o1 == outer[string]{box[string]{"k"}, 2})

	counts := map[box[string]]int{}
	counts[box[string]{"k"}]++
	counts[box[string]{"k"}]++
	counts[box[string]{"j"}]++
	fmt.Println(counts[box[string]{"k"}], counts[box[string]{"j"}], len(counts))
}

File: src/tests/Behavioral/GenericStructEquality/GenericStructEquality.cs
namespace go;

using fmt = fmt_package;

partial class main_package {

// Go runs an imported package's `init` before this package's own; .NET would never load
// an assembly nothing has touched yet, so that initialization is forced here.
[GoInit] internal static void initᴛᴛimportꓸfmt() {
    builtin.initPackage(typeof(fmt_package));
}

[GoType] partial struct handle<T>
    where T : /* comparable */ new()
{
    internal ж<T> p;
}

[GoType] partial struct box<T> {
    internal T v;
}

[GoType] partial struct meta<T> {
    internal nint hits;
}

[GoType] partial struct optval<T> {
    internal T v;
    internal bool valid;
}

[GoType] partial struct outer<T>
    where T : /* comparable */ new()
{
    internal box<T> inner;
    internal nint n;
}

internal static void Main() {
    ref var x = ref heap<nint>(out var Ꮡx);
    x = 5;
    ref var y = ref heap<nint>(out var Ꮡy);
    y = 5;
    var (h1, h2, h3) = (new handle<nint>(Ꮡx), new handle<nint>(Ꮡx), new handle<nint>(Ꮡy));
    fmt.Println(h1 == h2, h1 == h3, h1 == new handle<nint>(nil));
    var b1 = new box<@string>("go"u8);
    fmt.Println(b1 == new box<@string>("go"u8), b1 == new box<@string>("cs"u8));
    fmt.Println(new box<nint>(7) == new box<nint>(7), new box<nint>(7) == new box<nint>(8));
    var (m1, m2) = (new meta<@string>(3), new meta<float64>(3));
    fmt.Println(m1 == new meta<@string>(3), m2 == new meta<float64>(4));
    var (n1, n2) = (new optval<@string>("a"u8, true), new optval<@string>("a"u8, false));
    fmt.Println(n1 == new optval<@string>("a"u8, true), n1 == n2);
    var (o1, o2) = (new outer<@string>(new box<@string>("k"u8), 1), new outer<@string>(new box<@string>("k"u8), 1));
    fmt.Println(o1 == o2, o1 == new outer<@string>(new box<@string>("k"u8), 2));
    var counts = new map<box<@string>, nint>{};
    counts[new box<@string>("k"u8)]++;
    counts[new box<@string>("k"u8)]++;
    counts[new box<@string>("j"u8)]++;
    fmt.Println(counts[new box<@string>("k"u8)], counts[new box<@string>("j"u8)], len(counts));
}

} // end main_package
